Support for Evpn symmetric routing + EVPN Prefix route

mkanjari opened the pull request:
The PR largely includes the implementation of following two IETF drafts:
1. https://tools.ietf.org/html/draft-ietf-bess-evpn-inter-subnet-forwarding-03
2. https://tools.ietf.org/html/draft-ietf-bess-evpn-prefix-advertisement-08

donaldsharp commented:
@mkanjari - Please fix the subject lines in some of the above commits

louberger commented:
@mkanjari 
looks like just one valgrind loss:
==5151== HEAP SUMMARY:
==5151==     in use at exit: 8,192 bytes in 1 blocks
==5151==   total heap usage: 263,022 allocs, 263,021 frees, 14,978,681 bytes allocated
==5151== 
==5151== Searching for pointers to 1 not-freed blocks
==5151== Checked 411,824 bytes
==5151== 
==5151== 8,192 bytes in 1 blocks are definitely lost in loss record 1 of 1
==5151==    at 0x4A057BB: calloc (vg_replace_malloc.c:593)
==5151==    by 0x42C357: bgp_master_init (bgpd.c:7404)
==5151==    by 0x4200D6: main (bgp_main.c:380)
==5151== 
{
   <insert_a_suppression_name_here>
   Memcheck:Leak
   fun:calloc
   fun:bgp_master_init
   fun:main
}
==5151== LEAK SUMMARY:
==5151==    definitely lost: 8,192 bytes in 1 blocks
==5151==    indirectly lost: 0 bytes in 0 blocks
==5151==      possibly lost: 0 bytes in 0 blocks
==5151==    still reachable: 0 bytes in 0 blocks
==5151==         suppressed: 0 bytes in 0 blocks
==5151==

mkanjari commented on zebra/zebra_vty.c:
```diff
@@ -1875,6 +1884,80 @@ DEFUN (show_vrf,
+	assert(zvrf);
+
+	ret = zebra_vxlan_process_vrf_vni_cmd(zvrf, vni, err, 1);
+	if (ret != 0) {
```
sure, made the changes and uploaded a new diff.

mkanjari commented on zebra/zebra_vty.c:
```diff
@@ -1875,6 +1884,80 @@ DEFUN (show_vrf,
+		if (!zvrf)
+			continue;
+
+		vty_out(vty, "vrf: %s VNI: %u",
```
We cold just show the VRF which have a VNI mapped to them, will make the change.

mkanjari commented:
@pguibert6WIND Made suggested changes. 
Sorry the reference is wrong. We implemented the https://tools.ietf.org/html/draft-ietf-bess-evpn-prefix-advertisement-08. Updated the title in the PR. I will have to read them again to see the gap.

louberger commented:
@mkanjari FYI same error is being detected multiple times:

==3501== ERROR SUMMARY: 2 errors from 1 contexts (suppressed: 6 from 6)
==3501== 
==3501== 2 errors in context 1 of 1:
==3501== Invalid read of size 4
==3501==    at 0x49D649: bgp_evpn_cleanup (bgp_evpn.c:4358)
==3501==    by 0x427348: bgp_free (bgpd.c:3256)
==3501==    by 0x42AD14: bgp_delete (bgpd.c:3208)
==3501==    by 0x420DF0: sigint (bgp_main.c:183)
==3501==    by 0x4CF582: quagga_sigevent_process (sigevent.c:103)
==3501==    by 0x4D735D: thread_fetch (thread.c:1321)
==3501==    by 0x4C2116: frr_run (libfrr.c:878)
==3501==    by 0x420BAE: main (bgp_main.c:404)
==3501==  Address 0x4d095a0 is 0 bytes inside a block of size 8,192 free'd
==3501==    at 0x4A06430: free (vg_replace_malloc.c:446)
==3501==    by 0x421A62: bgp_terminate (bgpd.c:7567)
==3501==    by 0x420D9E: sigint (bgp_main.c:147)
==3501==    by 0x4CF582: quagga_sigevent_process (sigevent.c:103)
==3501==    by 0x4D735D: thread_fetch (thread.c:1321)
==3501==    by 0x4C2116: frr_run (libfrr.c:878)
==3501==    by 0x420BAE: main (bgp_main.c:404)
==3501== 
{
   <insert_a_suppression_name_here>
   Memcheck:Addr4
   fun:bgp_evpn_cleanup
   fun:bgp_free
   fun:bgp_delete
   fun:sigint
   fun:quagga_sigevent_process
   fun:thread_fetch
   fun:frr_run
   fun:main
}
--3501-- 
--3501-- used_suppression:      4 U1004-ARM-_dl_relocate_object
--3501-- used_suppression:      2 glibc-2.5.x-on-SUSE-10.2-(PPC)-2a
==3501== 
==3501== ERROR SUMMARY: 2 errors from 1 contexts (suppressed: 6 from 6)
-------------------------------------------------------------------------------
ERRORs from: /scratch/home/sw/regression/frr-reference/master/github/testconfigs/Regression/Logs/bgpd-2017-12-14-06:01:49.127.0.0.137.valgrind (val-error)
-------------------------------------------------------------------------------
==3525== ERROR SUMMARY: 2 errors from 1 contexts (suppressed: 6 from 6)
==3525== 
==3525== 2 errors in context 1 of 1:
==3525== Invalid read of size 4
==3525==    at 0x49D649: bgp_evpn_cleanup (bgp_evpn.c:4358)
==3525==    by 0x427348: bgp_free (bgpd.c:3256)
==3525==    by 0x42AD14: bgp_delete (bgpd.c:3208)
==3525==    by 0x420DF0: sigint (bgp_main.c:183)
==3525==    by 0x4CF582: quagga_sigevent_process (sigevent.c:103)
==3525==    by 0x4D735D: thread_fetch (thread.c:1321)
==3525==    by 0x4C2116: frr_run (libfrr.c:878)
==3525==    by 0x420BAE: main (bgp_main.c:404)
==3525==  Address 0x4d095a0 is 0 bytes inside a block of size 8,192 free'd
==3525==    at 0x4A06430: free (vg_replace_malloc.c:446)
==3525==    by 0x421A62: bgp_terminate (bgpd.c:7567)
==3525==    by 0x420D9E: sigint (bgp_main.c:147)
==3525==    by 0x4CF582: quagga_sigevent_process (sigevent.c:103)
==3525==    by 0x4D735D: thread_fetch (thread.c:1321)
==3525==    by 0x4C2116: frr_run (libfrr.c:878)
==3525==    by 0x420BAE: main (bgp_main.c:404)
==3525== 
{
   <insert_a_suppression_name_here>
   Memcheck:Addr4
   fun:bgp_evpn_cleanup
   fun:bgp_free
   fun:bgp_delete
   fun:sigint
   fun:quagga_sigevent_process
   fun:thread_fetch
   fun:frr_run
   fun:main
}
--3525-- 
--3525-- used_suppression:      4 U1004-ARM-_dl_relocate_object
--3525-- used_suppression:      2 glibc-2.5.x-on-SUSE-10.2-(PPC)-2a
==3525== 
==3525== ERROR SUMMARY: 2 errors from 1 contexts (suppressed: 6 from 6)

mkanjari commented:
@pguibert6WIND Addressed all the comments. Please let me know if you have any further suggestions, improvements. 
@louberger : Finally addressed the CI issues ! 
Thanks @donaldsharp for the help in changing the commit subject lines. I have addressed the SA issues as well. Please let me know if anything is missing.

mkanjari commented:
@louberger @pguibert6WIND @donaldsharp : addressed all the review comments.

donaldsharp commented:
I just spoke with @louberger and I believe he is satisfied with the code here.

pguibert6WIND commented:
Hi @mkanjari , all will be done by next tuesday at the next meeting.

bochi-michael commented on bgpd/bgp_route.c:
```diff
@@ -2219,6 +2219,14 @@ static void bgp_process_main_one(struct bgp *bgp, struct bgp_node *rn,
 		}
 	}
 
+	/* advertise/withdraw type-5 routes */
```
what's the purpose of this logic? in my test this causes frr to send type-5 routes received back to it's peer

pguibert6WIND commented on bgpd/bgpd.h:
```diff
@@ -408,6 +409,16 @@ struct bgp {
+
+	/* vrf flags */
+	uint32_t vrf_flags;
+#define BGP_VRF_AUTO (1 << 0)
```
maybe vrf_flags could have been  moved to a struct vrf * structure.

here this is just a remark. no need to move it now.

mkanjari replied:
I agree that we should have something simillar for lib/vrf as well. However, these particular flags are only specific to bgpd. bgpd has special handling for auto created VRF (VRF which is not yet created in the kernel).

pguibert6WIND commented on bgpd/bgp_evpn_private.h:
```diff
@@ -222,6 +222,15 @@ static inline vni_t label2vni(mpls_label_t *label)
+	eval->val[1] = ECOMMUNITY_EVPN_SUBTYPE_ROUTERMAC;
+	memcpy(&eval->val[2], rmac, ETH_ALEN);
+}
+
```
looks like a duplicate with bgp_attr_evpn.c, function bgp_add_routermac_ecom

mkanjari replied:
Yes seems like a duplicate. I guess one of the function was already there in frr. We should remove bgp_add_routermac_ecom. Other extended communities follow a simillar semantic to encode_rmac_extcomm. Please let me know your thoughts. 
However, this doesn't seem like exact duplicate one function is taking struct ecommunity_val as a parameter while other is taking struct attr *attr.

pguibert6WIND commented on bgpd/bgp_evpn.c:
```diff
@@ -604,8 +604,15 @@ static int bgp_zebra_send_remote_vtep(struct bgp *bgp, struct bgpevpn *vpn,
  * Build extended communities for EVPN route. RT and ENCAP are
  * applicable to all routes.
+ * TODO: currently kernel doesnt support ipv6 routes with ipv4 nexthops.
+ * This means that we can't do symmetric routing for ipv6 hosts routes
```
do you know if IPv4-Mapped IPv6 Addresses can be supported ?

mkanjari replied:
These will not be supported either. We have a plan to support them later with general ipv6 route support for EVPN routing.

pguibert6WIND commented on bgpd/bgp_evpn.c:
```diff
@@ -2282,13 +2285,14 @@ static int install_uninstall_route_in_vrfs(struct bgp *bgp_def, afi_t afi,
 
-	/* Only type-2 routes go into a VRF */
-	if (!(evp->prefix.route_type == BGP_EVPN_MAC_IP_ROUTE))
+	/* Only type-2/type-5 routes go into a VRF */
```
by the way, RT1 handling is not supported yet.

mkanjari replied:
yes, RT-1 is not yet supported. Processing for type-1 route not reach this point, it will dropped earlier. We will have to change this handling when we start supporting RT-1.

pguibert6WIND commented:
this is a good work done.
I have addresses some comments.
I just need some answers, ( it is related to functional perimeter)
but this is not blocking. it can be merged.

mkanjari commented:
Thanks @pguibert6WIND for the review. I have addressed all comments. Please let me know your thoughts. If nothing major is blocking, could you please merge the PR. I guess as the primary reviewer you have merge permission.
